src/support/hazard_weak.c

```diff
@@ -0,0 +1,277 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+#define WT_HAZARD_WEAK_FORALL(s, wha, whp)                     \
+    for (wha = (s)->hazard_weak; wha != NULL; wha = wha->next) \
+        for (whp = wha->hazard; whp < wha->hazard + wha->hazard_inuse; whp++)
+
+#define WT_HAZARD_WEAK_FORALL_BARRIER(s, wha, whp)               \
+    for (wha = (s)->hazard_weak; wha != NULL; wha = wha->next) { \
+        uint32_t __hazard_inuse;                                 \
+        WT_ORDERED_READ(__hazard_inuse, wha->hazard_inuse);      \
+        for (whp = wha->hazard; whp < wha->hazard + __hazard_inuse; whp++)
+
+#define WT_HAZARD_WEAK_FORALL_BARRIER_END }
+
+/*
+ * __wt_hazard_weak_close --
+ *     Verify that no weak hazard pointers are set.
+ */
+void
+__wt_hazard_weak_close(WT_SESSION_IMPL *session)
+{
+    WT_HAZARD_WEAK *whp;
+    WT_HAZARD_WEAK_ARRAY *wha;
+    uint32_t nhazard_weak;
+    bool found;
+
+    /*
+     * Check for a set weak hazard pointer and complain if we find one. We could just check the
+     * session's weak hazard pointer count, but this is a useful diagnostic.
+     */
+    for (found = false, nhazard_weak = 0, wha = session->hazard_weak; wha != NULL;
```

Comment:
My preference is to use `WT_HAZARD_WEAK_FORALL` here and move the initialization outside of the for loop for the non iterated variables.

Reply:
Here other than walking all the weak hazard slots, we are also calculating per array used slots (nhazard), which is done per array and not per hazard slot. i.e. in the first loop not in the second loop. This makes directly using the WT_HAZARD_WEAK_FORALL macro hard. Happy to a make change if you suggest an alternative that uses the macro.

Reply:
Ah okay happy to leave it as is then.